### PR TITLE
implemented TCX Huber-JPCRD-2016 for CO2

### LINKED ISSUE #2265: Sharp non-differentiable changes in thermal conductivity of CO2 and other gases
Sudden sharp changes leading to non-differentiable behavior in certain ranges of temperature for thermal conductivity are seen for gases like CO2, Methane and possibly others (not verified).

## The following is a plot for thermal conductivity of `CO2` at a pressure of `8MPa`.
image

## Zooming at around 304K shows this behavior explicitly.
image

## Comparision with the commercial RFPROP Library shows significant deviation of the properties as well as absence of such a behavior.
image

## Similar behavior is seen for `Methane` at `8MPa`
image

###  The problem can be assumed to be related to the underlying interpolation schemes or with the transition from one way of  calculation of the properties to another due to its presence in different gases.

### PR DESCRIPTION
**Pull request by Arnab-Mahanti**
### Description of the Change

 This change implements the new thermal conductivity model Huber-JPCRD-2016-CO2 for CO2. This changes the thermal conductivity contributions as follows:
- Dilute contribution: hardcoded `CarbonDioxideHuberJPCRD2016`
- Critical contribution: `simplified_Olchowy_Sengers` with values from Huber 2016
- Residual contribution: `polynomial` with values from Huber 2016 

### Benefits

[The older model from Scalabrin et. al. contains  a non-physical bump near the critical temperature of CO2 even though the density is far away from the critical point. Comparision with [NIST Thermophysical Properties]( https://webbook.nist.gov/chemistry/fluid/) shows this phenomenon to be non-existent.

### Possible Drawbacks

Not known

### Verification Process

Here is a comparison of the three models: CoolProp (current), REFPROP (Actually corresponds to NIST Thermophysical Properties), CoolProp (modified).

image

### Applicable Issues

Closes #2265

### CODE REVIEW

**Comment by ibell** [comment]
Thank you very much, this is very nice work. Some small comments in my review before I merge.

**Comment by Arnab-Mahanti** [comment]
@ibell Have you commented anything? Sorry, I didn't understand what you meant.

**Review by ibell — CHANGES_REQUESTED** [review]

**Comment by ibell** [comment]
Doh I forgot to close out the review

File: src/Backends/Helmholtz/TransportRoutines.cpp [hunk 1]
```diff
@@ -767,34 +767,12 @@ CoolPropDbl TransportRoutines::conductivity_critical_hardcoded_CO2_ScalabrinJPCR
 
 CoolPropDbl TransportRoutines::conductivity_dilute_hardcoded_CO2(HelmholtzEOSMixtureBackend& HEOS) {
```
[the remark this answers is not in the document]
Comment (Arnab-Mahanti) [reply src/Backends/Helmholtz/TransportRoutines.cpp:768]: done

File: CoolPropBibTeXLibrary.bib [hunk 2]
```diff
@@ -3258,4 +3258,13 @@ @Misc{IAPWS-SurfaceTension-1994
   Year                     = {1994}
 }
 
+@article{Huber-JPCRD-2016-CO2,
+	title = {{Reference Correlation of the Thermal Conductivity of Carbon Dioxide from the Triple Point to 1100 K and up to 200 MPa}},
+	volume = {45},
+	doi = {10.1063/1.4940892},
+	number = {1},
+	journal = {Journal of Physical and Chemical Reference Data},
+	author = {M.L. Huber and E.A. Sykioti and M.J. Assael and  R.A. Perkins},
```
[the remark this answers is not in the document]
Comment (Arnab-Mahanti) [reply CoolPropBibTeXLibrary.bib]: Done. 
`Note:` Many others have a similar issue.

**Comment by Arnab-Mahanti** [comment]
Ah wait. There are some issues with the commits.

**Comment by Arnab-Mahanti** [comment]
What a mess. I have added the changes. @ibell

**Comment by ibell** [comment]
Looks good I think. Can you confirm that you get the validation data from the paper with your code after all the changes? It would be best to update the points in the testing too, but that can be after.

**Review by ibell — APPROVED** [review]

**Comment by ibell** [comment]
Once this is merged, want to tackle the same thing for viscosity? You are a quick study of the CoolProp code

**Comment by Arnab-Mahanti** [comment]
I will try if I can find some time.

**Comment by Arnab-Mahanti** [comment]
I guess the viscosity model does not have a major issue for CO2. Following is a comparison at P=8e6.

image

**Comment by ibell** [comment]
No major issue, but it should be more accurate

**Comment by ibell** [comment]
So the last thing: please confirm you get the validation data in the paper for conductivity model

**Comment by Arnab-Mahanti** [comment]
### Validation with the reference paper

The following program was used for validation.

```c++
#include <iostream>
#include <CoolProp.h>
#include <matplot/matplot.h>
#include <iomanip>

int main(int argc, char const *argv[])
{
    auto [P, T] = matplot::meshgrid(matplot::linspace(20, 200, 10), matplot::linspace(300, 1100, 9));
    auto k = matplot::transform(P, T, [](auto p, auto t)
                                { return CoolProp::PropsSI("CONDUCTIVITY", "P", p * 1e6, "T", t, "CO2"); });

    for (auto &&row : k)
    {
        for (auto &&i : row)
        {
            std::cout << std::setprecision(4) << i * 1000 << "\t";
        }
        std::cout << "\n";
    }

    return 0;
}
```

The output for the range of `PXT = [20:10:200]X[300:1001100]` is as follows:

image

The values for `[0,0.1]X[240,300:100:1100]` are as follows:
image

For the values of 0MPa Pressure  Exceptions occur at the `HelmholtzEOSMixtureBackend::post_update` step called from `solver_rho_Tp`. Here is a view of the backtrace of the call stack.

image

The values obtained for 240K at P=[20:20:120] are as follows:
165.5   180.2   193.1   204.7   215.5   225.7

**Comment by Arnab-Mahanti** [comment]
Except for P=0MPa the values are in excellent agreement. A typo has been fixed.

**Comment by ibell** [comment]
Did you check with the tabulated check values in the paper? There should be like 7 of them

**Comment by Arnab-Mahanti** [comment]
These values are  from Table 8

**Comment by Arnab-Mahanti** [comment]
Similar phenomenon at \rho=0. Some problem in the solver of EOS.

image

For T=310K:
The comparison is done with automatic viscosity estimation only.

image

Critical enhancement is not implemented empirically, but rather through the simplified crossover model. (Section 3.3).

**Comment by ibell** [comment]
Right, usually I use a density of 1e-6 or something for the zero densities. Anyhow, thanks for the nice contribution, merging

**Pull request merged**